app/controllers/lti_deployments_controller.rb

```diff
@@ -28,75 +30,119 @@ def launch
   end
 
   def redirect_login
-    if params[:id_token].blank? || params[:state] != session.id.to_s
-      render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') }, layout: false
-      return
-    end
-    referrer_uri = URI(request.referer)
-    # Get LMS JWK set
-    jwk_url = "#{referrer_uri.scheme}://#{referrer_uri.host}:#{referrer_uri.port}#{self.class::LMS_JWK_ENDPOINT}"
-    # A list of public keys and associated metadata for JWTs signed by the LMS
-    lms_jwks = JSON.parse(Net::HTTP.get_response(URI(jwk_url)).body)
-    begin
-      decoded_token = JWT.decode(
-        params[:id_token], # Encoded JWT signed by LMS
-        nil, # If the token is passphrase-protected, set the passphrase here
-        true, # Verify the signature of this token
-        algorithms: ['RS256'],
-        iss: "#{referrer_uri.scheme}://#{referrer_uri.host}",
-        verify_iss: true,
-        aud: session[:client_id], # OpenID Connect uses client ID as the aud parameter
-        verify_aud: true,
-        jwks: lms_jwks # The correct JWK will be selected by matching jwk kid param with id_token kid
-      )
-    rescue JWT::DecodeError
-      render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') }, layout: false
-      return
-    end
-    lti_params = decoded_token[0]
-    unless lti_params['nonce'] == session[:nonce]
-      render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') }, layout: false
+    if request.post?
+      if params[:id_token].blank? || params[:state] != session.id.to_s
+        render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') },
+                                     layout: false
+        return
+      end
+      referrer_uri = URI(request.referer)
+      referrer_uri = referrer_uri.to_s.remove(referrer_uri.path) if referrer_uri.path
+      referrer_uri = referrer_uri.to_s.remove(URI(referrer_uri).query) if URI(referrer_uri).query
+      referrer_uri = referrer_uri.to_s.remove('?')
+      # Get LMS JWK set
+      jwk_url = "#{referrer_uri}#{self.class::LMS_JWK_ENDPOINT}"
+      # A list of public keys and associated metadata for JWTs signed by the LMS
+      lms_jwks = JSON.parse(Net::HTTP.get_response(URI(jwk_url)).body)
+      begin
+        decoded_token = JWT.decode(
+          params[:id_token], # Encoded JWT signed by LMS
+          nil, # If the token is passphrase-protected, set the passphrase here
+          true, # Verify the signature of this token
+          algorithms: ['RS256'],
+          iss: cookies.encrypted[:iss],
+          verify_iss: true,
+          aud: cookies.encrypted[:client_id], # OpenID Connect uses client ID as the aud parameter
+          verify_aud: true,
+          jwks: lms_jwks # The correct JWK will be selected by matching jwk kid param with id_token kid
+        )
+      rescue JWT::DecodeError
+        cookies.delete(:iss)
```

Comment:
can we put the cookie deletion in to an ensure block instead so that we can be sure they're always deleted whether the rest of the code succeeds or not